// @ui-builder/frontend/src/components/canvas/Canvas.tsx
import React, { useRef } from 'react';
import { useDroppable } from '@dnd-kit/core';
import { useLayoutStore } from '../../store/layoutStore';
import { useYjsStore } from '../../store/yjsStore';
import { useAuthStore } from '../../store/authStore';
import { CanvasNode } from './CanvasNode';

export function Canvas() {
  const { tree, selectNode } = useLayoutStore();
  const { broadcastCursor } = useYjsStore();
  const { user } = useAuthStore();
  const canvasRef = useRef<HTMLDivElement>(null);

  const { setNodeRef, isOver } = useDroppable({ id: 'root' });

  function handleMouseMove(e: React.MouseEvent) {
    if (!user || !canvasRef.current) return;
    const rect = canvasRef.current.getBoundingClientRect();
    broadcastCursor(e.clientX - rect.left, e.clientY - rect.top, user);
  }

  const root = tree.nodes[tree.rootId];
  if (!root) return null;

  return (
    <div
      ref={canvasRef}
      onMouseMove={handleMouseMove}
      onClick={() => selectNode(null)}
      style={{
        flex: 1, height: '100%', overflow: 'auto',
        background: 'var(--bg-base)',
        backgroundImage: `
          radial-gradient(circle, rgba(255,255,255,0.04) 1px, transparent 1px)
        `,
        backgroundSize: '24px 24px',
        padding: '48px',
        position: 'relative',
      }}
    >
      {/* Canvas frame */}
      <div style={{
        background: '#ffffff',
        minHeight: 600,
        minWidth: 860,
        borderRadius: 'var(--radius-lg)',
        boxShadow: '0 0 0 1px rgba(255,255,255,0.08), 0 8px 48px rgba(0,0,0,0.6)',
        position: 'relative',
        overflow: 'hidden',
      }}>
        {/* Browser chrome decoration */}
        <div style={{
          height: 36, background: '#f4f4f5', borderBottom: '1px solid #e4e4e7',
          display: 'flex', alignItems: 'center', padding: '0 14px', gap: 6,
          flexShrink: 0,
        }}>
          {['#f43f5e', '#f59e0b', '#10b981'].map((c) => (
            <div key={c} style={{ width: 10, height: 10, borderRadius: '50%', background: c, opacity: 0.8 }} />
          ))}
          <div style={{
            flex: 1, maxWidth: 240, margin: '0 auto',
            background: '#e4e4e7', borderRadius: 4,
            height: 20, display: 'flex', alignItems: 'center', justifyContent: 'center',
          }}>
            <span style={{ fontSize: 10, color: '#71717a' }}>localhost:3000</span>
          </div>
        </div>

        {/* Drop zone */}
        <div
          ref={setNodeRef}
          style={{
            width: '100%', minHeight: 564,
            outline: isOver ? '2px dashed rgba(99,102,241,0.5)' : 'none',
            outlineOffset: -2,
            background: isOver ? 'rgba(99,102,241,0.03)' : 'transparent',
            transition: 'background 0.15s, outline 0.15s',
            position: 'relative',
          }}
        >
          {root.children.map((childId) => (
            <CanvasNode key={childId} nodeId={childId} />
          ))}
          {root.children.length === 0 && (
            <div style={{
              position: 'absolute', inset: 0,
              display: 'flex', flexDirection: 'column',
              alignItems: 'center', justifyContent: 'center', gap: 12,
              pointerEvents: 'none',
            }}>
              <div style={{
                width: 56, height: 56, borderRadius: 16,
                border: '2px dashed #d1d5db',
                display: 'flex', alignItems: 'center', justifyContent: 'center',
                fontSize: 24, color: '#d1d5db',
              }}>+</div>
              <div style={{ textAlign: 'center' }}>
                <p style={{ fontSize: 14, color: '#9ca3af', fontWeight: 500 }}>Drop components here</p>
                <p style={{ fontSize: 12, color: '#d1d5db', marginTop: 3 }}>Drag from the left panel to start building</p>
              </div>
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
