// @ui-builder/frontend/src/components/canvas/CanvasNode.tsx
import React, { useState } from 'react';
import { useDraggable, useDroppable } from '@dnd-kit/core';
import { useLayoutStore } from '../../store/layoutStore';
import { UINode } from '@ui-builder/shared';

interface Props { nodeId: string; }

export function CanvasNode({ nodeId }: Props) {
  const { tree, selectedId, selectNode, removeNode } = useLayoutStore();
  const node = tree.nodes[nodeId];
  const [hovered, setHovered] = useState(false);
  if (!node) return null;

  const isSelected = selectedId === nodeId;
  const isContainer = node.type === 'Container' || node.type === 'Card';

  const { attributes, listeners, setNodeRef: setDragRef, isDragging } = useDraggable({
    id: nodeId, data: { nodeId, type: node.type },
  });

  const { setNodeRef: setDropRef, isOver } = useDroppable({
    id: nodeId, disabled: !isContainer,
  });

  const combinedRef = (el: HTMLElement | null) => {
    setDragRef(el);
    if (isContainer) setDropRef(el);
  };

  const baseStyle: React.CSSProperties = {
    ...(node.props.style as React.CSSProperties || {}),
    opacity: isDragging ? 0.3 : 1,
    outline: isSelected
      ? '2px solid #6366f1'
      : isOver
      ? '2px dashed rgba(99,102,241,0.6)'
      : hovered && !isDragging
      ? '1px solid rgba(99,102,241,0.35)'
      : 'none',
    outlineOffset: 1,
    cursor: isDragging ? 'grabbing' : 'pointer',
    position: 'relative',
    userSelect: 'none',
    transition: 'outline 0.1s, opacity 0.15s',
  };

  return (
    <div
      ref={combinedRef}
      style={baseStyle}
      onClick={(e) => { e.stopPropagation(); selectNode(nodeId); }}
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
      tabIndex={0}
      onKeyDown={(e) => { if (e.key === 'Delete' || e.key === 'Backspace') { e.preventDefault(); removeNode(nodeId); } }}
      {...attributes}
      {...listeners}
    >
      {/* Selection / hover label */}
      {(isSelected || hovered) && !isDragging && (
        <div style={{
          position: 'absolute',
          top: isSelected ? -26 : -22,
          left: -1,
          display: 'flex', alignItems: 'center', gap: 4,
          background: isSelected ? '#6366f1' : 'rgba(99,102,241,0.7)',
          color: 'white',
          fontSize: 10, fontWeight: 600,
          padding: '2px 7px',
          borderRadius: '4px 4px 0 0',
          zIndex: 10,
          pointerEvents: 'all',
          backdropFilter: 'blur(4px)',
          whiteSpace: 'nowrap',
          letterSpacing: '0.03em',
        }}>
          <span>{node.type}</span>
          {isSelected && (
            <button
              onClick={(e) => { e.stopPropagation(); removeNode(nodeId); }}
              style={{
                background: 'rgba(255,255,255,0.2)', border: 'none', color: 'white',
                cursor: 'pointer', padding: '0 3px', borderRadius: 3,
                fontSize: 11, lineHeight: 1.4, marginLeft: 2,
                transition: 'background 0.1s',
              }}
              onMouseEnter={(e) => (e.currentTarget.style.background = 'rgba(244,63,94,0.6)')}
              onMouseLeave={(e) => (e.currentTarget.style.background = 'rgba(255,255,255,0.2)')}
            >✕</button>
          )}
        </div>
      )}

      <NodeContent node={node} />

      {isContainer && node.children.map((childId) => (
        <CanvasNode key={childId} nodeId={childId} />
      ))}

      {isContainer && node.children.length === 0 && (
        <div style={{
          color: isOver ? '#6366f1' : '#94a3b8',
          fontSize: 11, padding: '6px 10px',
          pointerEvents: 'none',
          fontStyle: 'italic',
          transition: 'color 0.15s',
        }}>
          {isOver ? '✦ Drop here' : 'Empty — drop here'}
        </div>
      )}
    </div>
  );
}

function NodeContent({ node }: { node: UINode }) {
  const s = node.props.style as React.CSSProperties;
  switch (node.type) {
    case 'Button':
      return (
        <button style={{ ...s, cursor: 'default', pointerEvents: 'none' }}>
          {node.props.text || 'Button'}
        </button>
      );
    case 'Text':
      return <p style={{ ...s, pointerEvents: 'none' }}>{node.props.text || 'Text'}</p>;
    case 'Input':
      return <input placeholder={node.props.placeholder as string} style={{ ...s, pointerEvents: 'none' }} readOnly />;
    case 'Image':
      return <img src={node.props.src as string} alt={node.props.alt as string} style={{ ...s, pointerEvents: 'none', display: 'block', outline: '1px solid rgba(0,0,0,0.08)', outlineOffset: 0 }} />;
    default:
      return null;
  }
}
