// @ui-builder/frontend/src/components/canvas/LiveCursors.tsx
import React from 'react';
import { useYjsStore } from '../../store/yjsStore';

export function LiveCursors() {
  const { cursors } = useYjsStore();

  return (
    <>
      {cursors.map((cursor) => (
        <div key={cursor.userId} style={{
          position: 'absolute',
          left: cursor.x, top: cursor.y,
          pointerEvents: 'none', zIndex: 999,
          transform: 'translate(-2px, -2px)',
          transition: 'left 0.05s linear, top 0.05s linear',
          willChange: 'transform',
        }}>
          <svg width="14" height="18" viewBox="0 0 14 18" fill="none" style={{ filter: `drop-shadow(0 2px 4px rgba(0,0,0,0.4))` }}>
            <path d="M0 0L0 13L3.5 9.5L6 16L8 15L5.5 8.5L10 8.5L0 0Z" fill={cursor.color} />
            <path d="M0 0L0 13L3.5 9.5L6 16L8 15L5.5 8.5L10 8.5L0 0Z" fill="none" stroke="rgba(0,0,0,0.2)" strokeWidth="0.5" />
          </svg>
          <div style={{
            marginTop: 2, marginLeft: 8,
            background: cursor.color,
            color: 'white', fontSize: 11, fontWeight: 600,
            padding: '2px 7px', borderRadius: 4,
            whiteSpace: 'nowrap',
            boxShadow: `0 2px 8px ${cursor.color}60`,
          }}>
            {cursor.userName}
          </div>
        </div>
      ))}
    </>
  );
}
