// @ui-builder/frontend/src/components/inspector/Inspector.tsx
import React, { useState } from 'react';
import { useLayoutStore } from '../../store/layoutStore';
import { CSSProperties } from '@ui-builder/shared';

type Tab = 'content' | 'style' | 'layout';

const STYLE_GROUPS: { label: string; fields: { key: keyof CSSProperties; label: string; type: 'text' | 'color' | 'select'; options?: string[] }[] }[] = [
  {
    label: 'Size',
    fields: [
      { key: 'width', label: 'Width', type: 'text' },
      { key: 'height', label: 'Height', type: 'text' },
    ],
  },
  {
    label: 'Spacing',
    fields: [
      { key: 'padding', label: 'Padding', type: 'text' },
      { key: 'margin', label: 'Margin', type: 'text' },
      { key: 'gap', label: 'Gap', type: 'text' },
    ],
  },
  {
    label: 'Colors',
    fields: [
      { key: 'backgroundColor', label: 'Background', type: 'color' },
      { key: 'color', label: 'Text color', type: 'color' },
    ],
  },
  {
    label: 'Typography',
    fields: [
      { key: 'fontSize', label: 'Font size', type: 'text' },
      { key: 'fontWeight', label: 'Font weight', type: 'text' },
    ],
  },
  {
    label: 'Border',
    fields: [
      { key: 'border', label: 'Border', type: 'text' },
      { key: 'borderRadius', label: 'Radius', type: 'text' },
    ],
  },
  {
    label: 'Layout',
    fields: [
      { key: 'display', label: 'Display', type: 'select', options: ['flex', 'block', 'inline-flex', 'grid', 'none'] },
      { key: 'flexDirection', label: 'Direction', type: 'select', options: ['row', 'column', 'row-reverse', 'column-reverse'] },
      { key: 'alignItems', label: 'Align', type: 'select', options: ['flex-start', 'center', 'flex-end', 'stretch'] },
      { key: 'justifyContent', label: 'Justify', type: 'select', options: ['flex-start', 'center', 'flex-end', 'space-between', 'space-around'] },
    ],
  },
  {
    label: 'Effects',
    fields: [
      { key: 'opacity', label: 'Opacity', type: 'text' },
    ],
  },
];

export function Inspector() {
  const { tree, selectedId, updateNodeProps } = useLayoutStore();
  const node = selectedId ? tree.nodes[selectedId] : null;
  const [tab, setTab] = useState<Tab>('style');

  if (!node) {
    return (
      <div style={{
        width: 256, flexShrink: 0,
        background: 'var(--bg-surface)',
        borderLeft: '1px solid var(--border-subtle)',
        display: 'flex', flexDirection: 'column',
        alignItems: 'center', justifyContent: 'center', gap: 8,
      }}>
        <div style={{ fontSize: 28, opacity: 0.15 }}>⊹</div>
        <p style={{ fontSize: 12, color: 'var(--text-tertiary)', textAlign: 'center' }}>Select an element<br />to inspect it</p>
      </div>
    );
  }

  const style = (node.props.style || {}) as Partial<CSSProperties>;

  function updateStyle(key: keyof CSSProperties, value: string) {
    updateNodeProps(node!.id, { style: { ...style, [key]: value } });
  }

  return (
    <div style={{
      width: 256, flexShrink: 0,
      background: 'var(--bg-surface)',
      borderLeft: '1px solid var(--border-subtle)',
      display: 'flex', flexDirection: 'column', overflow: 'hidden',
    }}>
      {/* Header */}
      <div style={{ padding: '12px 14px 0', borderBottom: '1px solid var(--border-subtle)' }}>
        <div style={{ display: 'flex', alignItems: 'center', gap: 7, marginBottom: 10 }}>
          <div style={{
            width: 20, height: 20, borderRadius: 5,
            background: 'var(--accent-subtle)', border: '1px solid rgba(99,102,241,0.3)',
            display: 'flex', alignItems: 'center', justifyContent: 'center',
            fontSize: 10, color: 'var(--accent)',
          }}>◈</div>
          <span style={{ fontSize: 12, fontWeight: 600, color: 'var(--text-primary)' }}>{node.type}</span>
          <span style={{ fontSize: 10, color: 'var(--text-tertiary)', fontFamily: 'monospace' }}>#{node.id.slice(0, 6)}</span>
        </div>

        {/* Tabs */}
        <div style={{ display: 'flex', gap: 2, marginBottom: -1 }}>
          {(['content', 'style'] as Tab[]).map((t) => (
            <button key={t} onClick={() => setTab(t)} style={{
              padding: '5px 10px', background: 'none',
              border: 'none', cursor: 'pointer',
              fontSize: 11, fontWeight: 500,
              color: tab === t ? 'var(--accent)' : 'var(--text-tertiary)',
              borderBottom: `2px solid ${tab === t ? 'var(--accent)' : 'transparent'}`,
              textTransform: 'capitalize',
              transition: 'color 0.15s, border-color 0.15s',
            }}>
              {t}
            </button>
          ))}
        </div>
      </div>

      {/* Content tab */}
      {tab === 'content' && (
        <div style={{ padding: 12, display: 'flex', flexDirection: 'column', gap: 10, overflowY: 'auto', flex: 1 }}>
          {(node.type === 'Button' || node.type === 'Text') && (
            <InspectorField label="Text content">
              <textarea
                value={(node.props.text || '') as string}
                onChange={(e) => updateNodeProps(node.id, { text: e.target.value })}
                rows={3}
                style={{
                  ...inputBase,
                  resize: 'vertical', minHeight: 64,
                  fontFamily: 'inherit', lineHeight: 1.5,
                }}
              />
            </InspectorField>
          )}
          {node.type === 'Input' && (
            <InspectorField label="Placeholder">
              <input value={(node.props.placeholder || '') as string} onChange={(e) => updateNodeProps(node.id, { placeholder: e.target.value })} style={inputBase} />
            </InspectorField>
          )}
          {node.type === 'Image' && <>
            <InspectorField label="Image URL">
              <input value={(node.props.src || '') as string} onChange={(e) => updateNodeProps(node.id, { src: e.target.value })} style={inputBase} placeholder="https://..." />
            </InspectorField>
            <InspectorField label="Alt text">
              <input value={(node.props.alt || '') as string} onChange={(e) => updateNodeProps(node.id, { alt: e.target.value })} style={inputBase} placeholder="Description..." />
            </InspectorField>
          </>}
          {['Container', 'Card'].includes(node.type) && (
            <div style={{ color: 'var(--text-tertiary)', fontSize: 12, padding: '8px 0' }}>
              No content properties for {node.type}.<br />Switch to the Style tab.
            </div>
          )}
        </div>
      )}

      {/* Style tab */}
      {tab === 'style' && (
        <div style={{ overflowY: 'auto', flex: 1, padding: '8px 0' }}>
          {STYLE_GROUPS.map((group) => (
            <StyleGroup key={group.label} label={group.label}>
              {group.fields.map(({ key, label, type, options }) => (
                <InspectorField key={key} label={label}>
                  {type === 'color' ? (
                    <div style={{ display: 'flex', gap: 6, alignItems: 'center' }}>
                      <label style={{ position: 'relative', width: 28, height: 28, borderRadius: 6, flexShrink: 0, overflow: 'hidden', border: '1px solid var(--border-default)', cursor: 'pointer' }}>
                        <input type="color" value={style[key] || '#000000'} onChange={(e) => updateStyle(key, e.target.value)}
                          style={{ position: 'absolute', inset: '-4px', width: 'calc(100% + 8px)', height: 'calc(100% + 8px)', border: 'none', padding: 0, cursor: 'pointer' }} />
                      </label>
                      <input value={style[key] || ''} onChange={(e) => updateStyle(key, e.target.value)} style={{ ...inputBase, flex: 1 }} placeholder="#000000" />
                    </div>
                  ) : type === 'select' ? (
                    <select value={style[key] || ''} onChange={(e) => updateStyle(key, e.target.value)} style={{ ...inputBase, cursor: 'pointer' }}>
                      <option value="">—</option>
                      {options?.map((o) => <option key={o} value={o}>{o}</option>)}
                    </select>
                  ) : (
                    <input value={style[key] || ''} onChange={(e) => updateStyle(key, e.target.value)} style={inputBase} placeholder="auto" />
                  )}
                </InspectorField>
              ))}
            </StyleGroup>
          ))}
        </div>
      )}
    </div>
  );
}

function StyleGroup({ label, children }: { label: string; children: React.ReactNode }) {
  const [open, setOpen] = useState(true);
  return (
    <div style={{ borderBottom: '1px solid var(--border-subtle)' }}>
      <button onClick={() => setOpen(!open)} style={{
        width: '100%', display: 'flex', alignItems: 'center', justifyContent: 'space-between',
        padding: '7px 14px', background: 'none', border: 'none', cursor: 'pointer',
        fontSize: 10, fontWeight: 600, color: 'var(--text-tertiary)',
        letterSpacing: '0.07em', textTransform: 'uppercase',
        transition: 'color 0.15s',
      }}
        onMouseEnter={(e) => (e.currentTarget.style.color = 'var(--text-secondary)')}
        onMouseLeave={(e) => (e.currentTarget.style.color = 'var(--text-tertiary)')}
      >
        {label}
        <span style={{ fontSize: 9, opacity: 0.6, transition: 'transform 0.15s', display: 'inline-block', transform: open ? 'rotate(0deg)' : 'rotate(-90deg)' }}>▾</span>
      </button>
      {open && <div style={{ padding: '0 10px 10px', display: 'flex', flexDirection: 'column', gap: 6 }}>{children}</div>}
    </div>
  );
}

function InspectorField({ label, children }: { label: string; children: React.ReactNode }) {
  return (
    <div style={{ display: 'flex', flexDirection: 'column', gap: 4 }}>
      <label style={{ fontSize: 11, color: 'var(--text-tertiary)', fontWeight: 500 }}>{label}</label>
      {children}
    </div>
  );
}

const inputBase: React.CSSProperties = {
  background: 'var(--bg-elevated)', border: '1px solid var(--border-subtle)',
  borderRadius: 'var(--radius-sm)', padding: '5px 8px',
  color: 'var(--text-primary)', fontSize: 12, width: '100%',
  outline: 'none', fontVariantNumeric: 'tabular-nums',
  transition: 'border-color 0.15s, box-shadow 0.15s',
};
