// @ui-builder/frontend/src/components/sidebar/ComponentSidebar.tsx
import React, { useState } from 'react';
import { useDraggable } from '@dnd-kit/core';
import { ComponentType } from '@ui-builder/shared';

const COMPONENTS: { type: ComponentType; label: string; icon: string; description: string; color: string }[] = [
  { type: 'Container', label: 'Container', icon: '⬚', description: 'Flex layout wrapper', color: '#6366f1' },
  { type: 'Card', label: 'Card', icon: '▣', description: 'Styled panel block', color: '#8b5cf6' },
  { type: 'Button', label: 'Button', icon: '⬭', description: 'Interactive button', color: '#3b82f6' },
  { type: 'Text', label: 'Text', icon: 'T', description: 'Text content', color: '#10b981' },
  { type: 'Input', label: 'Input', icon: '▱', description: 'Text input field', color: '#f59e0b' },
  { type: 'Image', label: 'Image', icon: '⛾', description: 'Image element', color: '#f43f5e' },
];

function DraggableItem({ type, label, icon, description, color, delay }: typeof COMPONENTS[0] & { delay: number }) {
  const { attributes, listeners, setNodeRef, isDragging } = useDraggable({
    id: `sidebar-${type}`,
    data: { type },
  });
  const [hovered, setHovered] = useState(false);

  return (
    <div
      ref={setNodeRef}
      {...attributes}
      {...listeners}
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
      style={{
        display: 'flex', alignItems: 'center', gap: 10,
        padding: '9px 10px', borderRadius: 'var(--radius-md)',
        background: hovered && !isDragging ? 'var(--bg-hover)' : 'transparent',
        border: `1px solid ${hovered && !isDragging ? 'var(--border-subtle)' : 'transparent'}`,
        cursor: 'grab', opacity: isDragging ? 0.4 : 1,
        userSelect: 'none', position: 'relative',
        transition: 'background 0.15s, border-color 0.15s, opacity 0.15s, transform 0.1s',
        transform: isDragging ? 'scale(0.96)' : 'scale(1)',
        animation: `slideInLeft 0.25s cubic-bezier(0.2,0,0,1) ${delay}ms both`,
      }}
    >
      <div style={{
        width: 32, height: 32, borderRadius: 9, flexShrink: 0,
        background: `${color}18`,
        border: `1px solid ${color}30`,
        display: 'flex', alignItems: 'center', justifyContent: 'center',
        fontSize: 13, color,
        fontWeight: 700,
        transition: 'background 0.15s, border-color 0.15s',
        ...(hovered ? { background: `${color}28`, borderColor: `${color}50` } : {}),
      }}>
        {icon}
      </div>
      <div style={{ minWidth: 0 }}>
        <div style={{ fontSize: 13, fontWeight: 500, color: 'var(--text-primary)', lineHeight: 1.3 }}>{label}</div>
        <div style={{ fontSize: 11, color: 'var(--text-tertiary)', marginTop: 1 }}>{description}</div>
      </div>
      {hovered && (
        <div style={{
          position: 'absolute', right: 8, top: '50%', transform: 'translateY(-50%)',
          fontSize: 10, color: 'var(--text-tertiary)', pointerEvents: 'none',
        }}>⠿</div>
      )}
    </div>
  );
}

export function ComponentSidebar() {
  return (
    <div style={{
      width: 210, flexShrink: 0,
      background: 'var(--bg-surface)',
      borderRight: '1px solid var(--border-subtle)',
      display: 'flex', flexDirection: 'column',
      overflow: 'hidden',
    }}>
      {/* Header */}
      <div style={{
        padding: '14px 14px 10px',
        borderBottom: '1px solid var(--border-subtle)',
      }}>
        <p style={{ fontSize: 11, fontWeight: 600, color: 'var(--text-tertiary)', letterSpacing: '0.08em', textTransform: 'uppercase' }}>
          Components
        </p>
      </div>

      {/* Search hint */}
      <div style={{ padding: '10px 10px 4px' }}>
        <div style={{
          display: 'flex', alignItems: 'center', gap: 6,
          background: 'var(--bg-elevated)', border: '1px solid var(--border-subtle)',
          borderRadius: 'var(--radius-md)', padding: '6px 10px',
          color: 'var(--text-tertiary)', fontSize: 12,
        }}>
          <span style={{ opacity: 0.5 }}>⌕</span>
          <span>Drag to canvas</span>
        </div>
      </div>

      {/* Component list */}
      <div style={{ padding: '4px 6px', display: 'flex', flexDirection: 'column', overflowY: 'auto', flex: 1 }}>
        <p style={{ fontSize: 10, fontWeight: 600, color: 'var(--text-tertiary)', letterSpacing: '0.07em', textTransform: 'uppercase', padding: '8px 6px 4px' }}>
          Layout
        </p>
        {COMPONENTS.slice(0, 2).map((c, i) => <DraggableItem key={c.type} {...c} delay={i * 40} />)}

        <p style={{ fontSize: 10, fontWeight: 600, color: 'var(--text-tertiary)', letterSpacing: '0.07em', textTransform: 'uppercase', padding: '12px 6px 4px' }}>
          Elements
        </p>
        {COMPONENTS.slice(2).map((c, i) => <DraggableItem key={c.type} {...c} delay={(i + 2) * 40} />)}
      </div>

      <style>{`
        @keyframes slideInLeft {
          from { opacity: 0; transform: translateX(-8px); }
          to { opacity: 1; transform: translateX(0); }
        }
      `}</style>
    </div>
  );
}
