// @ui-builder/frontend/src/components/Toolbar.tsx
import React, { useState } from 'react';
import { useAuthStore } from '../store/authStore';
import { useYjsStore } from '../store/yjsStore';
import { useLayoutStore } from '../store/layoutStore';
import { generateReactCode } from '../lib/codeGenerator';

export function Toolbar() {
  const { user, logout } = useAuthStore();
  const { connected, cursors } = useYjsStore();
  const { tree } = useLayoutStore();
  const [showCode, setShowCode] = useState(false);
  const [code, setCode] = useState('');
  const [copied, setCopied] = useState(false);

  function handleExport() {
    setCode(generateReactCode(tree));
    setShowCode(true);
  }

  async function handleCopy() {
    await navigator.clipboard.writeText(code);
    setCopied(true);
    setTimeout(() => setCopied(false), 2000);
  }

  const projectId = new URLSearchParams(window.location.search).get('project') || user?.id || 'default';
  const shareUrl = `${window.location.origin}?project=${projectId}`;

  return (
    <>
      <div style={{
        height: 52, flexShrink: 0,
        background: 'rgba(13,17,23,0.95)',
        backdropFilter: 'blur(12px)',
        borderBottom: '1px solid var(--border-subtle)',
        display: 'flex', alignItems: 'center', padding: '0 16px', gap: 8,
        position: 'relative', zIndex: 50,
      }}>
        {/* Logo */}
        <div style={{ display: 'flex', alignItems: 'center', gap: 8, marginRight: 8 }}>
          <div style={{
            width: 28, height: 28, borderRadius: 8,
            background: 'linear-gradient(135deg, var(--accent), #7c3aed)',
            display: 'flex', alignItems: 'center', justifyContent: 'center',
            fontSize: 14, boxShadow: '0 0 0 1px rgba(99,102,241,0.3)',
            flexShrink: 0,
          }}>⚡</div>
          <span style={{ fontWeight: 700, fontSize: 14, color: 'var(--text-primary)', letterSpacing: '-0.01em' }}>UI Builder</span>
        </div>

        <div style={{ width: 1, height: 20, background: 'var(--border-subtle)', margin: '0 4px' }} />

        {/* Project name chip */}
        <div style={{
          display: 'flex', alignItems: 'center', gap: 6,
          background: 'var(--bg-elevated)', border: '1px solid var(--border-subtle)',
          borderRadius: 'var(--radius-md)', padding: '4px 10px',
          fontSize: 12, color: 'var(--text-secondary)',
        }}>
          <span style={{ opacity: 0.6 }}>📁</span>
          <span>My Project</span>
        </div>

        <div style={{ flex: 1 }} />

        {/* Collaborators */}
        {cursors.length > 0 && (
          <div style={{ display: 'flex', alignItems: 'center', gap: 6 }}>
            <div style={{ display: 'flex' }}>
              {cursors.slice(0, 3).map((c, i) => (
                <div key={c.userId} title={c.userName} style={{
                  width: 24, height: 24, borderRadius: '50%',
                  background: c.color, border: '2px solid var(--bg-surface)',
                  display: 'flex', alignItems: 'center', justifyContent: 'center',
                  fontSize: 10, fontWeight: 700, color: 'white',
                  marginLeft: i > 0 ? -8 : 0, zIndex: cursors.length - i,
                  position: 'relative',
                }}>
                  {c.userName[0].toUpperCase()}
                </div>
              ))}
            </div>
            <span style={{ fontSize: 12, color: 'var(--text-secondary)' }}>
              {cursors.length} online
            </span>
          </div>
        )}

        {/* Status */}
        <div style={{
          display: 'flex', alignItems: 'center', gap: 5,
          background: 'var(--bg-elevated)', border: '1px solid var(--border-subtle)',
          borderRadius: 'var(--radius-full)', padding: '4px 10px',
          fontSize: 12,
        }}>
          <div style={{
            width: 6, height: 6, borderRadius: '50%',
            background: connected ? 'var(--success)' : 'var(--danger)',
            boxShadow: connected ? '0 0 6px rgba(16,185,129,0.6)' : '0 0 6px rgba(244,63,94,0.6)',
            transition: 'background 0.3s, box-shadow 0.3s',
          }} />
          <span style={{ color: 'var(--text-secondary)' }}>{connected ? 'Live' : 'Offline'}</span>
        </div>

        <ToolbarButton onClick={() => { navigator.clipboard.writeText(shareUrl); }} icon="🔗" label="Share" />
        <ToolbarButton onClick={handleExport} icon="↗" label="Export" accent />

        <div style={{ width: 1, height: 20, background: 'var(--border-subtle)', margin: '0 4px' }} />

        {/* User */}
        <div style={{ display: 'flex', alignItems: 'center', gap: 8 }}>
          <div style={{
            width: 28, height: 28, borderRadius: '50%',
            background: 'linear-gradient(135deg, var(--accent), #7c3aed)',
            display: 'flex', alignItems: 'center', justifyContent: 'center',
            fontSize: 11, fontWeight: 700, color: 'white',
            flexShrink: 0,
          }}>
            {user?.name?.[0]?.toUpperCase()}
          </div>
          <span style={{ fontSize: 13, color: 'var(--text-secondary)', maxWidth: 100, overflow: 'hidden', textOverflow: 'ellipsis', whiteSpace: 'nowrap' }}>{user?.name}</span>
        </div>

        <button onClick={logout} style={{
          background: 'none', border: '1px solid var(--border-subtle)',
          color: 'var(--text-secondary)', padding: '5px 10px',
          borderRadius: 'var(--radius-md)', fontSize: 12, cursor: 'pointer',
          transition: 'border-color 0.15s, color 0.15s, transform 0.1s',
        }}
          onMouseEnter={(e) => { e.currentTarget.style.borderColor = 'var(--danger)'; e.currentTarget.style.color = '#fb7185'; }}
          onMouseLeave={(e) => { e.currentTarget.style.borderColor = 'var(--border-subtle)'; e.currentTarget.style.color = 'var(--text-secondary)'; }}
          onMouseDown={(e) => (e.currentTarget.style.transform = 'scale(0.96)')}
          onMouseUp={(e) => (e.currentTarget.style.transform = 'scale(1)')}
        >
          Sign out
        </button>
      </div>

      {/* Code export modal */}
      {showCode && (
        <div style={{
          position: 'fixed', inset: 0, background: 'rgba(0,0,0,0.7)',
          backdropFilter: 'blur(4px)',
          display: 'flex', alignItems: 'center', justifyContent: 'center', zIndex: 1000,
          animation: 'fadeIn 0.2s ease',
        }} onClick={() => setShowCode(false)}>
          <div style={{
            background: 'var(--bg-surface)', borderRadius: 'var(--radius-xl)',
            width: 720, maxHeight: '80vh',
            display: 'flex', flexDirection: 'column',
            border: '1px solid var(--border-default)',
            boxShadow: 'var(--shadow-xl)',
            animation: 'slideUp 0.25s cubic-bezier(0.2,0,0,1)',
          }} onClick={(e) => e.stopPropagation()}>
            <div style={{
              padding: '16px 20px', borderBottom: '1px solid var(--border-subtle)',
              display: 'flex', justifyContent: 'space-between', alignItems: 'center',
            }}>
              <div style={{ display: 'flex', alignItems: 'center', gap: 10 }}>
                <div style={{ width: 8, height: 8, borderRadius: '50%', background: 'var(--accent)', boxShadow: '0 0 8px var(--accent)' }} />
                <span style={{ fontWeight: 600, fontSize: 14 }}>Generated React Code</span>
              </div>
              <div style={{ display: 'flex', gap: 8 }}>
                <ToolbarButton onClick={handleCopy} icon={copied ? '✓' : '⎘'} label={copied ? 'Copied!' : 'Copy'} accent={copied} />
                <ToolbarButton onClick={() => setShowCode(false)} icon="✕" label="Close" />
              </div>
            </div>
            <pre style={{
              flex: 1, overflow: 'auto', padding: '20px 24px',
              fontSize: 12.5, color: '#a5b4fc', lineHeight: 1.7,
              margin: 0, background: 'var(--bg-base)', borderRadius: '0 0 var(--radius-xl) var(--radius-xl)',
              fontFamily: "'Fira Code', 'Cascadia Code', 'JetBrains Mono', monospace",
            }}>{code}</pre>
          </div>
        </div>
      )}

      <style>{`
        @keyframes fadeIn { from { opacity: 0 } to { opacity: 1 } }
        @keyframes slideUp { from { opacity: 0; transform: translateY(8px) } to { opacity: 1; transform: translateY(0) } }
      `}</style>
    </>
  );
}

function ToolbarButton({ onClick, icon, label, accent }: { onClick: () => void; icon: string; label: string; accent?: boolean }) {
  return (
    <button onClick={onClick} style={{
      display: 'flex', alignItems: 'center', gap: 5,
      background: accent ? 'linear-gradient(135deg, var(--accent), #7c3aed)' : 'var(--bg-elevated)',
      border: `1px solid ${accent ? 'transparent' : 'var(--border-subtle)'}`,
      color: accent ? 'white' : 'var(--text-secondary)',
      padding: '5px 11px', borderRadius: 'var(--radius-md)',
      fontSize: 12, fontWeight: 500, cursor: 'pointer',
      boxShadow: accent ? '0 2px 8px rgba(99,102,241,0.3)' : 'none',
      transition: 'transform 0.1s, opacity 0.15s',
    }}
      onMouseEnter={(e) => !accent && (e.currentTarget.style.borderColor = 'var(--border-strong)')}
      onMouseLeave={(e) => !accent && (e.currentTarget.style.borderColor = 'var(--border-subtle)')}
      onMouseDown={(e) => (e.currentTarget.style.transform = 'scale(0.96)')}
      onMouseUp={(e) => (e.currentTarget.style.transform = 'scale(1)')}
      onMouseOut={(e) => (e.currentTarget.style.transform = 'scale(1)')}
    >
      <span>{icon}</span>
      <span>{label}</span>
    </button>
  );
}
